tools/internal/autoupdate/registry.go

```diff
@@ -0,0 +1,348 @@
+	}
+
+	// Case 2: Handle images with a registry but no namespace like "k8s.gcr.io/pause"
+	if len(splittedImage) == 2 && strings.Contains(splittedImage[0], ".") {
+		return splittedImage[0], "", splittedImage[1]
+	}
+
+	// Case 3: Handle images with long paths like "gcr.io/cloud-provider-vsphere/csi/release/syncer"
+	if len(splittedImage) > 3 && strings.Contains(splittedImage[0], ".") {
+		return splittedImage[0], splittedImage[1], strings.Join(splittedImage[2:], "/")
+	}
+
+	// Default Case: Handle standard 3-part images like "quay.io/skopeo/stable"
+	return splittedImage[0], splittedImage[1], splittedImage[2]
+}
+
+func getSuseAuthToken(namespace, repository string) (string, error) {
+	tokenURL := "https://registry.suse.com/auth?service=SUSE+Linux+Docker+Registry"
+	req, _ := http.NewRequest("GET", tokenURL, nil)
+	params := req.URL.Query()
+	params.Add("scope", fmt.Sprintf("repository:%s/%s:pull", namespace, repository))
+	req.URL.RawQuery = params.Encode()
+
+	resp, err := doRequestWithRetries(req)
+	if err != nil {
+		return "", fmt.Errorf("suse token request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	body, _ := io.ReadAll(resp.Body)
+	var tokenData SuseTokenResponse
+	if err := json.Unmarshal(body, &tokenData); err != nil {
+		return "", fmt.Errorf("failed to unmarshal suse token: %w", err)
+	}
+	return tokenData.AccessToken, nil
+}
+
+func doRequestWithRetries(req *http.Request) (*http.Response, error) {
+	maxRetries := 10
+	backoffFactor := 1 * time.Second
+	retryableStatuses := map[int]bool{
+		http.StatusInternalServerError: true, // 500
+		http.StatusBadGateway:          true, // 502
+		http.StatusServiceUnavailable:  true, // 503
+		http.StatusGatewayTimeout:      true, // 504
+	}
+
+	var resp *http.Response
+	var err error
+
+	for i := 0; i < maxRetries; i++ {
+		resp, err = httpClient.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("request failed: %w", err)
+		}
+
+		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			return resp, nil
+		}
+
+		if !retryableStatuses[resp.StatusCode] {
+			return nil, fmt.Errorf("request failed with status %s", resp.Status)
+		}
+
+		if resp.Body != nil {
+			io.Copy(io.Discard, resp.Body)
+			resp.Body.Close()
+		}
+
+		if i == maxRetries-1 {
+			break
+		}
+
+		sleepDuration := backoffFactor * time.Duration(math.Pow(2, float64(i)))
+		time.Sleep(sleepDuration)
+	}
+
+	return nil, fmt.Errorf("request failed after %d retries", maxRetries)
+}
+
+// parseLinkHeader extracts the next URL from the Link header for pagination.
+// Example: </v2/epinio/epinio-server/tags/list?last=v1.10.0-rc1-5-g0d7c9121&n=200>; rel="next"
+func parseLinkHeader(linkHeader string) string {
+	if linkHeader == "" {
+		return ""
+	}
+	link := strings.TrimPrefix(linkHeader, "<")
+	link = strings.TrimSuffix(link, `>; rel="next"`)
+	return link
+}
+
+func (r *Registry) processRegistryResponse(body []byte, linkHeader, githubToken, registry, namespace, pkg string, page int) ([]string, error) {
```

Comment:
The duplicated `switch-case` statements in `processRegistryResponse` and `buildRequest` are code smells. Arguably the best way to organize this is to give each registry its own type, and implement the code in each case inside that type.

On the other hand, this code is largely a translation of the original python. If you're feeling motivated, maybe consider restructuring this. Otherwise, feel free to disregard this comment.